# SkillFactory
## Введение в ML, введение в sklearn

В этом задании мы с вами рассмотрим данные с конкурса [Задача предсказания отклика клиентов ОТП Банка](http://www.machinelearning.ru/wiki/index.php?title=%D0%97%D0%B0%D0%B4%D0%B0%D1%87%D0%B0_%D0%BF%D1%80%D0%B5%D0%B4%D1%81%D0%BA%D0%B0%D0%B7%D0%B0%D0%BD%D0%B8%D1%8F_%D0%BE%D1%82%D0%BA%D0%BB%D0%B8%D0%BA%D0%B0_%D0%BA%D0%BB%D0%B8%D0%B5%D0%BD%D1%82%D0%BE%D0%B2_%D0%9E%D0%A2%D0%9F_%D0%91%D0%B0%D0%BD%D0%BA%D0%B0_%28%D0%BA%D0%BE%D0%BD%D0%BA%D1%83%D1%80%D1%81%29)

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

plt.style.use('ggplot')
plt.rcParams['figure.figsize'] = (12,5)

### Грузим данные

Считаем описание данных

In [2]:

df_descr = pd.read_csv('data/otp_description.csv', sep='\t', encoding='utf8')

In [3]:
df_descr

ПОЛЕ  \
0                 AGREEMENT_RK   
1                          AGE   
2            SOCSTATUS_WORK_FL   
3            SOCSTATUS_PENS_FL   
4                       GENDER   
5                  CHILD_TOTAL   
6                   DEPENDANTS   
7                    EDUCATION   
8               MARITAL_STATUS   
9                 GEN_INDUSTRY   
10                   GEN_TITLE   
11                ORG_TP_STATE   
12             ORG_TP_FCAPITAL   
13                     JOB_DIR   
14               FAMILY_INCOME   
15             PERSONAL_INCOME   
16        REG_ADDRESS_PROVINCE   
17       FACT_ADDRESS_PROVINCE   
18     POSTAL_ADDRESS_PROVINCE   
19                 TP_PROVINCE   
20                   REGION_NM   
21                 REG_FACT_FL   
22                FACT_POST_FL   
23                 REG_POST_FL   
24            REG_FACT_POST_FL   
25         REG_FACT_POST_TP_FL   
26              FL_PRESENCE_FL   
27                    OWN_AUTO   
28                 AUTO_RUS_FL   
29              HS_PRESENCE_FL   
30             COT_PRESENCE_FL   
31             GAR_PRESENCE_FL   
32            LAND_PRESENCE_FL   
33                      CREDIT   
34                        TERM   
35                 FST_PAYMENT   
36              DL_DOCUMENT_FL   
37             GPF_DOCUMENT_FL   
38            FACT_LIVING_TERM   
39                   WORK_TIME   
40               FACT_PHONE_FL   
41                REG_PHONE_FL   
42                GEN_PHONE_FL   
43              LOAN_NUM_TOTAL   
44             LOAN_NUM_CLOSED   
45               LOAN_NUM_PAYM   
46                LOAN_DLQ_NUM   
47                LOAN_MAX_DLQ   
48            LOAN_AVG_DLQ_AMT   
49            LOAN_MAX_DLQ_AMT   
50  PREVIOUS_CARD_NUM_UTILIZED   

                                             ОПИСАНИЕ  
0          уникальный идентификатор объекта в выборке  
1                                     возраст клиента  
2   социальный статус клиента относительно работы ...  
3   социальный статус клиента относительно пенсии ...  
4                 пол клиента (1- женщины, 0-мужчины)  
5                            количество детей клиента  
6                       количество иждивенцев клиента  
7                                         образование  
8                                  семейное положение  
9                              отрасль работы клиента  
10                                          должность  
11                       форма собственности компании  
12                  отношение к иностранному капиталу  
13          направление деятельности в нутри компании  
14               семейный доход (несколько категорий)  
15                    личный доход клиента (в рублях)  
16                        область регистрации клиента  
17            область фактического пребывания клиента  
18                             почтовый адрес область  
19  область торговой точки, где клиент брал послед...  
20                                          регион РФ  
21  адрес регистрации и адрес фактического пребыва...  
22  адрес фактического пребывания клиента и его по...  
23  адрес регистрации клиента и его почтовый адрес...  
24  почтовый, фактический и адрес регистрации совп...  
25  область регистрации, фактического пребывания, ...  
26  наличие в собственности квартиры (1 - есть, 0 ...  
27                кол-во автомобилей в собственности   
28  наличие в собственности автомобиля российского...  
29  наличие в собственности загородного дома (1 - ...  
30  наличие в собственности котеджа (1 - есть, 0 -...  
31  наличие в собственности гаража (1 - есть, 0 - ...  
32  наличие в собственности земельного участка (1 ...  
33        сумма последнего кредита клиента (в рублях)  
34                                       срок кредита  
35                    первоначальный взнос (в рублях)  
36  в анкете клиент указал водительское удостовере...  
37  в анкете клиен указал ГПФ (1 - указал, 0 - не ...  
38  количество месяцев проживания по месту фактиче...  
39          время

Считаем обучающую выборки и тестовую (которую мы как бы не видим)

In [4]:
df_train = pd.read_csv('data/otp_train.csv', sep='\t', encoding='utf8')

In [5]:
df_train.shape

(15223, 52)

In [6]:
df_test = pd.read_csv('data/otp_test.csv', sep='\t', encoding='utf8')

In [7]:
df_test.shape

(14910, 52)

In [8]:
df_train.head()

AGREEMENT_RK  TARGET  AGE  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  GENDER  \
0      59910150       0   49                  1                  0       1   
1      59910230       0   32                  1                  0       1   
2      59910525       0   52                  1                  0       1   
3      59910803       0   39                  1                  0       1   
4      59911781       0   30                  1                  0       0   

   CHILD_TOTAL  DEPENDANTS            EDUCATION  MARITAL_STATUS  \
0            2           1  Среднее специальное  Состою в браке   
1            3           3              Среднее  Состою в браке   
2            4           0     Неполное среднее  Состою в браке   
3            1           1               Высшее  Состою в браке   
4            0           0              Среднее  Состою в браке   

             ...             REG_PHONE_FL GEN_PHONE_FL LOAN_NUM_TOTAL  \
0            ...                        0            1              1   
1            ...                        0            1              1   
2            ...                        0            1              2   
3            ...                        1            1              1   
4            ...                        0            1              2   

  LOAN_NUM_CLOSED LOAN_NUM_PAYM LOAN_DLQ_NUM LOAN_MAX_DLQ  LOAN_AVG_DLQ_AMT  \
0               1             6            2            1              1580   
1               1             6            1            1              4020   
2               1            11            0            0                 0   
3               1             6            3            1  1589,92333333333   
4               1            16            2            1           1152,15   

  LOAN_MAX_DLQ_AMT PREVIOUS_CARD_NUM_UTILIZED  
0             1580                        NaN  
1             4020                        NaN  
2                0                        NaN  
3             1590                        NaN  
4             2230                        NaN  

[5 rows x 52 columns]

## Объединим две выборки

Так как пока мы пока не умеем работать sklearn  Pipeline, то для того, чтобы после предобработки столбцы в двух выборках находились на своих местах.

Для того, чтобы в дальнейшем отделить их введем новый столбец "sample"

In [9]:
df_train.loc[:, 'sample'] = 'train'
df_test.loc[:, 'sample'] = 'test'

In [10]:
df = df_test.append(df_train).reset_index(drop=True)

In [11]:
df.shape

(30133, 53)

### Чуть-чуть посмотрим на данные

Посмотрим типы данных и их заполняемость

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30133 entries, 0 to 30132
Data columns (total 53 columns):
AGREEMENT_RK                  30133 non-null int64
TARGET                        30133 non-null int64
AGE                           30133 non-null int64
SOCSTATUS_WORK_FL             30133 non-null int64
SOCSTATUS_PENS_FL             30133 non-null int64
GENDER                        30133 non-null int64
CHILD_TOTAL                   30133 non-null int64
DEPENDANTS                    30133 non-null int64
EDUCATION                     30133 non-null object
MARITAL_STATUS                30133 non-null object
GEN_INDUSTRY                  27420 non-null object
GEN_TITLE                     27420 non-null object
ORG_TP_STATE                  27420 non-null object
ORG_TP_FCAPITAL               27425 non-null object
JOB_DIR                       27420 non-null object
FAMILY_INCOME                 30133 non-null object
PERSONAL_INCOME               30133 non-null object
REG_ADDRESS_PRO

Видим, что часть данных - object, скорее всего стоки.


Давайте выведем эти значения для каждого столбца

In [13]:
for i in df_train.columns: # перебираем все столбцы
    if str(df_train[i].dtype) == 'object': # если тип столбца - object
        print('='*10)
        print(i) # выводим название столбца
        print(set(df_train[i])) # выводим все его значения (но делаем set - чтоб значения не повторялись)
        print('\n') # выводим пустую строку

EDUCATION
{'Неполное среднее', 'Высшее', 'Среднее специальное', 'Ученая степень', 'Неоконченное высшее', 'Два и более высших образования', 'Среднее'}


MARITAL_STATUS
{'Состою в браке', 'Вдовец/Вдова', 'Гражданский брак', 'Не состоял в браке', 'Разведен(а)'}


GEN_INDUSTRY
{nan, 'Энергетика', 'Транспорт', 'Информационные технологии', 'Коммунальное хоз-во/Дорожные службы', 'Развлечения/Искусство', 'Сборочные производства', 'Химия/Парфюмерия/Фармацевтика', 'ЧОП/Детективная д-ть', 'Сельское хозяйство', 'Туризм', 'Управляющая компания', 'Нефтегазовая промышленность', 'Маркетинг', 'Недвижимость', 'СМИ/Реклама/PR-агенства', 'Салоны красоты и здоровья', 'Наука', 'Юридические услуги/нотариальные услуги', 'Страхование', 'Другие сферы', 'Строительство', 'Торговля', 'Логистика', 'Банк/Финансы', 'Государственная служба', 'Подбор персонала', 'Информационные услуги', 'Образование', 'Здравоохранение', 'Ресторанный бизнес/Общественное питание', 'Металлургия/Промышленность/Машиностроение'}


GEN_TITLE


In [14]:
for i in df.columns: # перебираем все столбцы
    if str(df[i].dtype) == 'object': # если тип столбца - object
        print('='*10)
        print(i) # выводим название столбца
        print(set(df[i])) # выводим все его значения (но делаем set - чтоб значения не повторялись)
        print('\n') # выводим пустую строку

EDUCATION
{'Неполное среднее', 'Высшее', 'Среднее специальное', 'Ученая степень', 'Неоконченное высшее', 'Два и более высших образования', 'Среднее'}


MARITAL_STATUS
{'Состою в браке', 'Гражданский брак', 'Вдовец/Вдова', 'Не состоял в браке', 'Разведен(а)'}


GEN_INDUSTRY
{nan, 'Энергетика', 'Транспорт', 'Коммунальное хоз-во/Дорожные службы', 'Информационные технологии', 'Развлечения/Искусство', 'Сборочные производства', 'Химия/Парфюмерия/Фармацевтика', 'ЧОП/Детективная д-ть', 'Сельское хозяйство', 'Туризм', 'Управляющая компания', 'Нефтегазовая промышленность', 'Маркетинг', 'Недвижимость', 'СМИ/Реклама/PR-агенства', 'Салоны красоты и здоровья', 'Наука', 'Юридические услуги/нотариальные услуги', 'Страхование', 'Другие сферы', 'Строительство', 'Торговля', 'Логистика', 'Банк/Финансы', 'Государственная служба', 'Подбор персонала', 'Информационные услуги', 'Образование', 'Здравоохранение', 'Ресторанный бизнес/Общественное питание', 'Металлургия/Промышленность/Машиностроение'}


GEN_TITLE


Mожно заметить что некоторые переменные, которые обозначены как строки (например PERSONAL_INCOME) на самом деле числа, но по какой-то причине были распознаны как строки

Причина же что использовалась запятая для разделения не целой части числа..

Перекодировать их можно например так:

In [15]:
#df['PERSONAL_INCOME'].map(lambda x: x.replace(',', '.')).astype('float')

In [16]:
df['PERSONAL_INCOME'] = df['PERSONAL_INCOME'].map(lambda x: x.replace(',', '.')).astype('float')

In [17]:
df['CREDIT'] = df['CREDIT'].map(lambda x: x.replace(',', '.')).astype('float')

In [18]:
df['FST_PAYMENT'] = df['FST_PAYMENT'].map(lambda x: x.replace(',', '.')).astype('float')

In [19]:
df['LOAN_MAX_DLQ_AMT'] = df['LOAN_MAX_DLQ_AMT'].map(lambda x: x.replace(',', '.')).astype('float')

In [20]:
df['LOAN_AVG_DLQ_AMT'] = df['LOAN_AVG_DLQ_AMT'].map(lambda x: x.replace(',', '.')).astype('float')

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30133 entries, 0 to 30132
Data columns (total 53 columns):
AGREEMENT_RK                  30133 non-null int64
TARGET                        30133 non-null int64
AGE                           30133 non-null int64
SOCSTATUS_WORK_FL             30133 non-null int64
SOCSTATUS_PENS_FL             30133 non-null int64
GENDER                        30133 non-null int64
CHILD_TOTAL                   30133 non-null int64
DEPENDANTS                    30133 non-null int64
EDUCATION                     30133 non-null object
MARITAL_STATUS                30133 non-null object
GEN_INDUSTRY                  27420 non-null object
GEN_TITLE                     27420 non-null object
ORG_TP_STATE                  27420 non-null object
ORG_TP_FCAPITAL               27425 non-null object
JOB_DIR                       27420 non-null object
FAMILY_INCOME                 30133 non-null object
PERSONAL_INCOME               30133 non-null float64
REG_ADDRESS_PR

In [22]:
#df = pd.get_dummies(df)

In [23]:
#df = pd.get_dummies(df, columns=
#               ['EDUCATION', 'MARITAL_STATUS', 
#                'FAMILY_INCOME', 'REG_ADDRESS_PROVINCE', 
#                'FACT_ADDRESS_PROVINCE', 
#                'POSTAL_ADDRESS_PROVINCE', 'REGION_NM', 
#                'sample']
#              )

In [24]:
#list(df.columns.values)

Такой эффект наблюдается в столбцах `PERSONAL_INCOME`, `CREDIT`, `FST_PAYMENT`, `LOAN_AVG_DLQ_AMT`, `LOAN_MAX_DLQ_AMT`

### Теперь ваше небольшое исследование

#### Задание 1. Есть ли пропуски в данных? Что с ними сделать?

(единственного верного ответа нет - аргументируйте)

не хватает данных тут:

REGION_NM                     30131 non-null object
GEN_INDUSTRY                  27420 non-null object
GEN_TITLE                     27420 non-null object
ORG_TP_STATE                  27420 non-null object
ORG_TP_FCAPITAL               27425 non-null object
JOB_DIR                       27420 non-null object
TP_PROVINCE                   29543 non-null object
эти данные мы обработает с помощью df = pd.get_dummies(df)


WORK_TIME                     27416 non-null float64
PREVIOUS_CARD_NUM_UTILIZED    600 non-null float64

для PREVIOUS_CARD_NUM_UTILIZED там где nan сделаем 0
в случае WORK_TIME подставим среднюю

но сделаем это позже.
###df = df.drop(['PREVIOUS_CARD_NUM_UTILIZED'], axis=1)
#df["WORK_TIME"].median()
48.0
#df['WORK_TIME'] = df['WORK_TIME'].fillna(48)
#df.info(verbose=True, null_counts=True) ##будет видно все данные

#### Задание 2. Есть ли категориальные признаки? Что с ними делать?

In [25]:
# есть, позже приведем к адекватному виду
df.select_dtypes(exclude=['object']).join(pd.get_dummies(df.select_dtypes(include=['object'])))

AGREEMENT_RK  TARGET  AGE  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  \
0          59909969       0   30                  1                  0   
1          59910420       0   48                  0                  1   
2          59911223       0   35                  1                  0   
3          59911322       0   22                  1                  0   
4          59911438       0   27                  1                  0   
5          59911769       0   44                  0                  1   
6          59912044       0   57                  1                  0   
7          59912382       0   33                  1                  0   
8          59912422       0   53                  0                  1   
9          59912485       1   33                  1                  0   
10         59912527       0   33                  1                  0   
11         59912542       0   40                  1                  0   
12         59912574       0   59                  0                  1   
13         59912645       0   39                  1                  0   
14         59912670       0   23                  1                  0   
15         59912684       0   28                  1                  0   
16         59912731       0   55                  1                  0   
17         59912931       0   57                  1                  1   
18         59912951       0   34                  1                  0   
19         59913256       0   51                  1                  0   
20         59913432       0   65                  0                  1   
21         59913524       1   39                  1                  0   
22         59913601       0   55                  1                  0   
23         59913650       1   33                  1                  0   
24         59913678       0   46                  1                  0   
25         59913739       0   40                  1                  0   
26         59913764       1   47                  1                  0   
27         59914013       0   23                  1                  0   
28         59914189       0   63                  0                  1   
29         59914336       0   23                  1                  0   
...             ...     ...  ...                ...                ...   
30103      75288570       0   36                  1                  0   
30104      75288759       0   49                  1                  0   
30105      75288846       0   51                  1                  1   
30106      75288884       0   64                  0                  1   
30107      75288922       0   61                  0                  1   
30108      75288976       0   49                  1                  0   
30109      75289030       0   28                  1                  0   
30110      75289115       1   34                  1                  0   
30111      75289165       0   56                  1                  1   
30112      75289228       0   34                  1                  0   
30113      75289331       0   40                  1                  0   
30114      75289366       1   43                  1                  0   
30115      75289474       0   33                  1                  0   
30116      75289649       0   56                  1                  0   
30117      75289804       0   25                  1                  0   
30118      75289810       0   37                  1                  0   
30119      75289867       1   48                  1                  0   
30120      75289907       0   30                  1                  0   
30121      75290193       0   35                  1                  0   
30122      75290209       0   55                  1                  0   
30123      75290347       1   36                  1                  0   
30124      75290611       0   31                  1                  0   
30125      

In [26]:
#df.corr()

In [27]:
#from pandas.plotting import scatter_matrix
#scatter_matrix(df, alpha=1, figsize=(10, 10))

#### Задание 3. Фунция предобработки

Напишите функцию, которая бы

* Удаляло идентификатор `AGREEMENT_RK`
* Избавлялась от проблем с '.' и ',' в стобцах PERSONAL_INCOME, CREDIT, FST_PAYMENT, LOAN_AVG_DLQ_AMT, LOAN_MAX_DLQ_AMT
* Что-то делала с пропусками
* Кодировала категориальные признаки

В результате, ваш датафрейм должен содержать только числа и не содержать пропусков!

In [47]:
def preproc_data(df_input):
    df_output = df_input.copy()
    
    
    ## Your Code Here
    # delete AGREEMENT_RK
    df_output = df_output.drop(['AGREEMENT_RK'], axis=1)
    
    #float to int
    df_output['PERSONAL_INCOME'] = df_output['PERSONAL_INCOME'].astype('int64')
    df_output['CREDIT'] = df_output['CREDIT'].astype('int64')
    df_output['FST_PAYMENT'] = df_output['FST_PAYMENT'].astype('int64')
    df_output['LOAN_AVG_DLQ_AMT'] = df_output['LOAN_AVG_DLQ_AMT'].astype('int64')
    df_output['LOAN_MAX_DLQ_AMT'] = df_output['LOAN_MAX_DLQ_AMT'].astype('int64')
    
    #work time fill
    df_output['WORK_TIME'] = df_output['WORK_TIME'].fillna(df_output["WORK_TIME"].median()).astype('int64')
    
    #prev card
    df_output['PREVIOUS_CARD_NUM_UTILIZED'] = df_output['PREVIOUS_CARD_NUM_UTILIZED'].fillna(0).astype('int64')
    
    # category code
    df_output = pd.get_dummies(df_output, columns=['EDUCATION', 'MARITAL_STATUS', 
                                                   'GEN_INDUSTRY', 'GEN_TITLE', 
                                                   'ORG_TP_STATE', 'ORG_TP_FCAPITAL', 
                                                   'JOB_DIR', 'FAMILY_INCOME', 
                                                   'REG_ADDRESS_PROVINCE', 
                                                   'FACT_ADDRESS_PROVINCE', 
                                                   'POSTAL_ADDRESS_PROVINCE', 
                                                   'TP_PROVINCE', 'REGION_NM'])
        
    return df_output

In [48]:
df_preproc = df.pipe(preproc_data)

df_train_preproc = df_preproc.query('sample == "train"').drop(['sample'], axis=1)
df_test_preproc = df_preproc.query('sample == "test"').drop(['sample'], axis=1)

In [49]:
df_preproc.head(5).T

0     1      2      3      4
TARGET                               0     0      0      0      0
AGE                                 30    48     35     22     27
SOCSTATUS_WORK_FL                    1     0      1      1      1
SOCSTATUS_PENS_FL                    0     1      0      0      0
GENDER                               1     1      0      0      1
CHILD_TOTAL                          1     0      1      0      2
DEPENDANTS                           1     0      0      0      2
PERSONAL_INCOME                   7000  4100  40000  20000  17000
REG_FACT_FL                          0     1      1      1      1
FACT_POST_FL                         1     1      1      1      1
REG_POST_FL                          0     1      1      1      1
REG_FACT_POST_FL                     0     1      1      1      1
REG_FACT_POST_TP_FL                  1     1      1      1      1
FL_PRESENCE_FL                       1     0      0      1      0
OWN_AUTO                             0     0      1      0      0
AUTO_RUS_FL                          0     0      1      0      0
HS_PRESENCE_FL                       0     1      0      0      0
COT_PRESENCE_FL                      0     0      0      0      0
GAR_PRESENCE_FL                      0     1      0      0      0
LAND_PRESENCE_FL                     0     0      0      0      0
CREDIT                            3016  6000  10000   5175  13065
TERM                                 5     6      6      4      6
FST_PAYMENT                        335  2000   2490    575  10000
DL_DOCUMENT_FL                       0     0      0      0      0
GPF_DOCUMENT_FL                      0     1      0      0      0
FACT_LIVING_TERM                    96   299     19    252     36
WORK_TIME                           39    48    156      9     84
FACT_PHONE_FL                        1     0      1      1      0
REG_PHONE_FL                         0     0      0      0      0
GEN_PHONE_FL                         1     0      1      1      1
...                                ...   ...    ...    ...    ...
TP_PROVINCE_Сахалинская область      0     0      0      0      0
TP_PROVINCE_Свердловская область     0     0      0      0      0
TP_PROVINCE_Смоленская область       0     0      0      0      0
TP_PROVINCE_Ставропольский край      0     0      0      0      0
TP_PROVINCE_Тамбовская область       0     0      0      0      0
TP_PROVINCE_Татарстан                0     0      0      0      0
TP_PROVINCE_Тверская область         0     0      0      0      0
TP_PROVINCE_Томская область          0     0      0      0      0
TP_PROVINCE_Тульская область         0     0      0      0      0
TP_PROVINCE_Тюменская область        0     0      0      0      0
TP_PROVINCE_Удмуртия                 0     0      0      0      0
TP_PROVINCE_Ульяновская область      0     0      0      0      0
TP_PROVINCE_Хабаровский край         0     0      0      0      1
TP_PROVINCE_Ханты-Мансийский АО      0     0      0      0      0
TP_PROVINCE_Челябинская область      0     0      0      0      0
TP_PROVINCE_Читинская область        0     0      0      1      0
TP_PROVINCE_Чувашия                  0     0      0      0      0
TP_PROVINCE_Якутия                   0     0      0      0      0
TP_PROVINCE_Ярославская область      0     0      0      0      0
REGION_NM_ВОСТОЧНО-СИБИРСКИЙ         1     0      0      1      0
REGION_NM_ДАЛЬНЕВОСТОЧНЫЙ            0     0      0      0      1
REGION_NM_ЗАПАДНО-СИБИРСКИЙ          0     0      0      0      0
REGION_NM_ПОВОЛЖСКИЙ                 0     0      0      0      0
REGION_NM_ПРИВОЛЖСКИЙ                0     0      0      0      0
REGION_NM_СЕВЕРО-ЗАПАДНЫЙ            0     0      0      0      0
REGION_NM_УРАЛЬСКИЙ                  0     0      0      0      0
REGION_NM_ЦЕНТРАЛЬНЫЙ 1              0     1      0      0      0
REGION_NM_ЦЕНТРАЛЬНЫЙ 2              0     0      1      0      0
REGION_NM_ЦЕНТРАЛЬНЫЙ ОФИС           0     0      0      0      0
REGION_NM_Ю

In [51]:
df_preproc.info(verbose=True, null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30133 entries, 0 to 30132
Data columns (total 444 columns):
TARGET                                                    30133 non-null int64
AGE                                                       30133 non-null int64
SOCSTATUS_WORK_FL                                         30133 non-null int64
SOCSTATUS_PENS_FL                                         30133 non-null int64
GENDER                                                    30133 non-null int64
CHILD_TOTAL                                               30133 non-null int64
DEPENDANTS                                                30133 non-null int64
PERSONAL_INCOME                                           30133 non-null int64
REG_FACT_FL                                               30133 non-null int64
FACT_POST_FL                                              30133 non-null int64
REG_POST_FL                                               30133 non-null int64
REG_FACT_POST_FL     

#### Задание 4. Отделите целевую переменную и остальные признаки

Должно получится:
* 2 матрицы: X и X_test
* 2 вектора: y и y_test

In [56]:
X = df_train_preproc.drop(['TARGET'], axis=1)
X_test = df_test_preproc.drop(['TARGET'], axis=1)

y = df_train_preproc['TARGET']
y_test = df_test_preproc['TARGET']

In [55]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier()
rfc.fit(X_train, y_train)
predict = rfc.predict(X_test)

print('precision', precision_score(y_test, predict))
print('recall', recall_score(y_test, predict))
print('accuracy', accuracy_score(y_test, predict))
print('f1', f1_score(y_test, predict))

NameError: name 'X_train' is not defined

In [54]:
from sklearn.tree import DecisionTreeClassifier
dtc = DecisionTreeClassifier()
x = X
y = X_test
dtc.fit(x, y)
predict = dtc.predict(df_train_preproc)

ValueError: Number of labels=14910 does not match number of samples=15223

#### Задание 5. Обучение и оценка качества разных моделей

In [58]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.1, random_state=42)

dtc = DecisionTreeClassifier(random_state=42)
dtc.fit(X, y)
predict = dtc.predict(X_test)

accuracy_score(y_test, predict)

NameError: name 'accuracy_score' is not defined

In [57]:
from sklearn.cross_validation import train_test_split
# test_size=0.3, random_state=42

## Your Code Here
X_train, X_test, y_train, y_test = train_test_split(
    X,
    X_test,
    y,
    y_test,
    test_size=0.3, random_state=42)



D:\Anaconda\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


ValueError: Found input variables with inconsistent numbers of samples: [15223, 14910, 15223, 14910]

In [46]:
# Попробовать следующие "черные ящики": интерфейс одинаковый 
#     fit, 
#     predict, 
#     predict_proba

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

## Your Code Here

In [48]:
# Посчитать метрики стандартные
# accuracy, precision, recall

from sklearn.metrics import accuracy_score, precision_score, recall_score

## Your Code Here

In [49]:
accuracy_score(y, predict)

NameError: name 'y' is not defined

In [ ]:
# Визуалищировать эти метрики всех моделей на одном графике (чтоб визуально посмотреть)

## Your Code Here

In [ ]:
# Потроить roc-кривые всех можелей на одном графике
# Вывести roc_auc каждой моделе

## Your Code Here

In [ ]:
from sklearn.cross_validation import cross_val_score
from sklearn.model_selection import StratifiedKFold
# Сделать k-fold (10 фолдов) кросс-валидацию каждой модели
# И посчитать средний roc_auc
cv = StratifiedKFold(n_splits=10, shuffle=True, random_state=123)

## Your Code Here

In [ ]:
# Взять лучшую модель и сделать predict (с вероятностями (!!!)) для test выборки


## Your Code Here

In [ ]:
# Померить roc_auc на тесте
